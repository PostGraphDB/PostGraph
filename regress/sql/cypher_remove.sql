/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

LOAD 'postgraph';
SET search_path TO postgraph;

CREATE GRAPH cypher_remove;
USE GRAPH cypher_remove;

--test 1
CREATE (:test_1);
CREATE (:test_1 {i: 0, j: 5, a: 0});
CREATE (:test_1 {i: 1});

MATCH (n:test_1) REMOVE n.i ;
MATCH (n:test_1) RETURN n;

--test 2
CREATE (:test_2);
CREATE (:test_2 {i: 0, j: 5, a: 0});
CREATE (:test_2 {i: 1});

MATCH (n:test_2) REMOVE n.j RETURN n;
MATCH (n:test_2) RETURN n;

--test 3 Validate Paths are updated
CREATE (:test_3 { i : 20 } )-[:test_3_edge {j:20}]->(:test_3 {i:10});

MATCH p=(n)-[:test_3_edge]->() REMOVE n.i RETURN p;

--test 4 Edges
MATCH (n) REMOVE n.i RETURN n;
CREATE (:test_4 { i : 20 } )-[:test_4_edge {j:20}]->(:test_4 {i:10});

MATCH ()-[n]->(:test_4) REMOVE n.i RETURN n;
MATCH ()-[n]->(:test_4) RETURN n;

--test 5 two REMOVE clauses
CREATE (:test_5 {i: 1, j : 2, k : 3}) ;

MATCH (n:test_5)
REMOVE n.i
REMOVE n.j
RETURN n
;


MATCH (n:test_5)
RETURN n
;

--test 6 Create a loop and see that set can work after create
CREATE (:test_6 {j: 5, y: 99});

MATCH (n {j: 5})
CREATE p=(n)-[e:e {j:34}]->(n)
REMOVE n.y
RETURN n, p
;
MATCH (n:test_6) RETURN n;


--test 7 Create a loop and see that set can work after create
CREATE (:test_7)-[e:e {j:34}]->()
REMOVE e.y
RETURN e
;

MATCH (n:test_7) RETURN n;



--test 8
MATCH (n:test_7)
MATCH (n)-[e:e {j:34}]->()
REMOVE n.y
RETURN n
;

--Handle Inheritance
CREATE ( {i : 1 });
MATCH (n) REMOVE n.i RETURN n;
MATCH (n) RETURN n;

-- prepared statements
/*
CREATE ( {i : 1 });
PREPARE p_1 AS MATCH (n) REMOVE n.i RETURN n ;
EXECUTE p_1;

CREATE ( {i : 1 });
EXECUTE p_1;
-- pl/pgsql
CREATE ( {i : 1 });

CREATE FUNCTION remove_test()
RETURNS TABLE(vertex vertex)
LANGUAGE plpgsql
VOLATILE
AS $BODY$
BEGIN
	RETURN QUERY MATCH (n) REMOVE n.i RETURN n;
END
$BODY$;

SELECT remove_test();

CREATE ( {i : 1 });
SELECT remove_test();

*/
--
-- Updating Multiple Fields
--
MATCH (n) RETURN n;
MATCH (n) REMOVE n.i, n.j, n.k RETURN n;
MATCH (n) RETURN n;

CREATE ()-[:edge_multi_property { i: 5, j: 20}]->();
MATCH ()-[e:edge_multi_property]-() RETURN e;
MATCH ()-[e:edge_multi_property]-() REMOVE e.i, e.j RETURN e;

--Errors
REMOVE n.i;

MATCH (n) REMOVE n.i = NULL;

MATCH (n) REMOVE wrong_var.i;

--
-- Clean up
--
DROP FUNCTION remove_test;
DROP GRAPH cypher_remove CASCADE;

--
-- End
--
