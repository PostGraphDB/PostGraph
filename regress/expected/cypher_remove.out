/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
LOAD 'postgraph';
SET search_path TO postgraph;
CREATE GRAPH cypher_remove;
NOTICE:  graph "cypher_remove" has been created
 create_graph 
--------------
 
(1 row)

USE GRAPH cypher_remove;
 use_graph 
-----------
 
(1 row)

--test 1
CREATE (:test_1);
--
(0 rows)

CREATE (:test_1 {i: 0, j: 5, a: 0});
--
(0 rows)

CREATE (:test_1 {i: 1});
--
(0 rows)

MATCH (n:test_1) REMOVE n.i ;
--
(0 rows)

MATCH (n:test_1) RETURN n;
                                     n                                      
----------------------------------------------------------------------------
 {"id": 844424930131969, "label": "test_1", "properties": {}}
 {"id": 844424930131970, "label": "test_1", "properties": {"a": 0, "j": 5}}
 {"id": 844424930131971, "label": "test_1", "properties": {}}
(3 rows)

--test 2
CREATE (:test_2);
--
(0 rows)

CREATE (:test_2 {i: 0, j: 5, a: 0});
--
(0 rows)

CREATE (:test_2 {i: 1});
--
(0 rows)

MATCH (n:test_2) REMOVE n.j RETURN n;
                                      n                                      
-----------------------------------------------------------------------------
 {"id": 1125899906842625, "label": "test_2", "properties": {}}
 {"id": 1125899906842626, "label": "test_2", "properties": {"a": 0, "i": 0}}
 {"id": 1125899906842627, "label": "test_2", "properties": {"i": 1}}
(3 rows)

MATCH (n:test_2) RETURN n;
                                      n                                      
-----------------------------------------------------------------------------
 {"id": 1125899906842625, "label": "test_2", "properties": {}}
 {"id": 1125899906842626, "label": "test_2", "properties": {"a": 0, "i": 0}}
 {"id": 1125899906842627, "label": "test_2", "properties": {"i": 1}}
(3 rows)

--test 3 Validate Paths are updated
CREATE (:test_3 { i : 20 } )-[:test_3_edge {j:20}]->(:test_3 {i:10});
--
(0 rows)

MATCH p=(n)-[:test_3_edge]->() REMOVE n.i RETURN p;
                                                                                                                                         p                                                                                                                                         
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 [{"id": 1407374883553281, "label": "test_3", "properties": {"i": 20}}, {"id": 1688849860263937, "start_id": 1407374883553281, "end_id": 1407374883553282, "label": "test_3_edge", "properties": {"j": 20}}, {"id": 1407374883553282, "label": "test_3", "properties": {"i": 10}}]
(1 row)

--test 4 Edges
MATCH (n) REMOVE n.i RETURN n;
                                     n                                      
----------------------------------------------------------------------------
 {"id": 844424930131969, "label": "test_1", "properties": {}}
 {"id": 844424930131970, "label": "test_1", "properties": {"a": 0, "j": 5}}
 {"id": 844424930131971, "label": "test_1", "properties": {}}
 {"id": 1125899906842625, "label": "test_2", "properties": {}}
 {"id": 1125899906842626, "label": "test_2", "properties": {"a": 0}}
 {"id": 1125899906842627, "label": "test_2", "properties": {}}
 {"id": 1407374883553282, "label": "test_3", "properties": {}}
 {"id": 1407374883553281, "label": "test_3", "properties": {}}
(8 rows)

CREATE (:test_4 { i : 20 } )-[:test_4_edge {j:20}]->(:test_4 {i:10});
--
(0 rows)

MATCH ()-[n]->(:test_4) REMOVE n.i RETURN n;
ERROR:  operator does not exist: label_id postgraph.= integer
HINT:  No operator matches the given name and argument types. You might need to add explicit type casts.
MATCH ()-[n]->(:test_4) RETURN n;
ERROR:  operator does not exist: label_id postgraph.= integer
HINT:  No operator matches the given name and argument types. You might need to add explicit type casts.
--test 5 two REMOVE clauses
CREATE (:test_5 {i: 1, j : 2, k : 3}) ;
--
(0 rows)

MATCH (n:test_5)
REMOVE n.i
REMOVE n.j
RETURN n
;
                                  n                                  
---------------------------------------------------------------------
 {"id": 2533274790395905, "label": "test_5", "properties": {"k": 3}}
(1 row)

MATCH (n:test_5)
RETURN n
;
                                  n                                  
---------------------------------------------------------------------
 {"id": 2533274790395905, "label": "test_5", "properties": {"k": 3}}
(1 row)

--test 6 Create a loop and see that set can work after create
CREATE (:test_6 {j: 5, y: 99});
--
(0 rows)

MATCH (n {j: 5})
CREATE p=(n)-[e:e {j:34}]->(n)
REMOVE n.y
RETURN n, p
;
                                     n                                      |                                                                                                                                            p                                                                                                                                            
----------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 {"id": 844424930131970, "label": "test_1", "properties": {"a": 0, "j": 5}} | [{"id": 844424930131970, "label": "test_1", "properties": {"a": 0, "j": 5}}, {"id": 3096224743817217, "start_id": 844424930131970, "end_id": 844424930131970, "label": "e", "properties": {"j": 34}}, {"id": 844424930131970, "label": "test_1", "properties": {"a": 0, "j": 5}}]
 {"id": 2814749767106561, "label": "test_6", "properties": {"j": 5}}        | [{"id": 2814749767106561, "label": "test_6", "properties": {"j": 5, "y": 99}}, {"id": 3096224743817218, "start_id": 2814749767106561, "end_id": 2814749767106561, "label": "e", "properties": {"j": 34}}, {"id": 2814749767106561, "label": "test_6", "properties": {"j": 5, "y": 99}}]
(2 rows)

MATCH (n:test_6) RETURN n;
                                  n                                  
---------------------------------------------------------------------
 {"id": 2814749767106561, "label": "test_6", "properties": {"j": 5}}
(1 row)

--test 7 Create a loop and see that set can work after create
CREATE (:test_7)-[e:e {j:34}]->()
REMOVE e.y
RETURN e
;
                                                            e                                                             
--------------------------------------------------------------------------------------------------------------------------
 {"id": 3096224743817219, "start_id": 3377699720527873, "end_id": 281474976710657, "label": "e", "properties": {"j": 34}}
(1 row)

MATCH (n:test_7) RETURN n;
                               n                               
---------------------------------------------------------------
 {"id": 3377699720527873, "label": "test_7", "properties": {}}
(1 row)

--test 8
MATCH (n:test_7)
MATCH (n)-[e:e {j:34}]->()
REMOVE n.y
RETURN n
;
                               n                               
---------------------------------------------------------------
 {"id": 3377699720527873, "label": "test_7", "properties": {}}
(1 row)

--Handle Inheritance
CREATE ( {i : 1 });
--
(0 rows)

MATCH (n) REMOVE n.i RETURN n;
                                     n                                      
----------------------------------------------------------------------------
 {"id": 281474976710657, "label": "", "properties": {}}
 {"id": 281474976710658, "label": "", "properties": {}}
 {"id": 844424930131969, "label": "test_1", "properties": {}}
 {"id": 844424930131971, "label": "test_1", "properties": {}}
 {"id": 844424930131970, "label": "test_1", "properties": {"a": 0, "j": 5}}
 {"id": 1125899906842625, "label": "test_2", "properties": {}}
 {"id": 1125899906842626, "label": "test_2", "properties": {"a": 0}}
 {"id": 1125899906842627, "label": "test_2", "properties": {}}
 {"id": 1407374883553282, "label": "test_3", "properties": {}}
 {"id": 1407374883553281, "label": "test_3", "properties": {}}
 {"id": 1970324836974593, "label": "test_4", "properties": {}}
 {"id": 1970324836974594, "label": "test_4", "properties": {}}
 {"id": 2533274790395905, "label": "test_5", "properties": {"k": 3}}
 {"id": 2814749767106561, "label": "test_6", "properties": {"j": 5}}
 {"id": 3377699720527873, "label": "test_7", "properties": {}}
(15 rows)

MATCH (n) RETURN n;
                                     n                                      
----------------------------------------------------------------------------
 {"id": 281474976710657, "label": "", "properties": {}}
 {"id": 281474976710658, "label": "", "properties": {}}
 {"id": 844424930131969, "label": "test_1", "properties": {}}
 {"id": 844424930131971, "label": "test_1", "properties": {}}
 {"id": 844424930131970, "label": "test_1", "properties": {"a": 0, "j": 5}}
 {"id": 1125899906842625, "label": "test_2", "properties": {}}
 {"id": 1125899906842626, "label": "test_2", "properties": {"a": 0}}
 {"id": 1125899906842627, "label": "test_2", "properties": {}}
 {"id": 1407374883553282, "label": "test_3", "properties": {}}
 {"id": 1407374883553281, "label": "test_3", "properties": {}}
 {"id": 1970324836974593, "label": "test_4", "properties": {}}
 {"id": 1970324836974594, "label": "test_4", "properties": {}}
 {"id": 2533274790395905, "label": "test_5", "properties": {"k": 3}}
 {"id": 2814749767106561, "label": "test_6", "properties": {"j": 5}}
 {"id": 3377699720527873, "label": "test_7", "properties": {}}
(15 rows)

-- prepared statements
/*
CREATE ( {i : 1 });
PREPARE p_1 AS MATCH (n) REMOVE n.i RETURN n ;
EXECUTE p_1;

CREATE ( {i : 1 });
EXECUTE p_1;
-- pl/pgsql
CREATE ( {i : 1 });

CREATE FUNCTION remove_test()
RETURNS TABLE(vertex vertex)
LANGUAGE plpgsql
VOLATILE
AS $BODY$
BEGIN
	RETURN QUERY MATCH (n) REMOVE n.i RETURN n;
END
$BODY$;

SELECT remove_test();

CREATE ( {i : 1 });
SELECT remove_test();

*/
--
-- Updating Multiple Fields
--
MATCH (n) RETURN n;
                                     n                                      
----------------------------------------------------------------------------
 {"id": 281474976710657, "label": "", "properties": {}}
 {"id": 281474976710658, "label": "", "properties": {}}
 {"id": 844424930131969, "label": "test_1", "properties": {}}
 {"id": 844424930131971, "label": "test_1", "properties": {}}
 {"id": 844424930131970, "label": "test_1", "properties": {"a": 0, "j": 5}}
 {"id": 1125899906842625, "label": "test_2", "properties": {}}
 {"id": 1125899906842626, "label": "test_2", "properties": {"a": 0}}
 {"id": 1125899906842627, "label": "test_2", "properties": {}}
 {"id": 1407374883553282, "label": "test_3", "properties": {}}
 {"id": 1407374883553281, "label": "test_3", "properties": {}}
 {"id": 1970324836974593, "label": "test_4", "properties": {}}
 {"id": 1970324836974594, "label": "test_4", "properties": {}}
 {"id": 2533274790395905, "label": "test_5", "properties": {"k": 3}}
 {"id": 2814749767106561, "label": "test_6", "properties": {"j": 5}}
 {"id": 3377699720527873, "label": "test_7", "properties": {}}
(15 rows)

MATCH (n) REMOVE n.i, n.j, n.k RETURN n;
                                  n                                  
---------------------------------------------------------------------
 {"id": 281474976710657, "label": "", "properties": {}}
 {"id": 281474976710658, "label": "", "properties": {}}
 {"id": 844424930131969, "label": "test_1", "properties": {}}
 {"id": 844424930131971, "label": "test_1", "properties": {}}
 {"id": 844424930131970, "label": "test_1", "properties": {"a": 0}}
 {"id": 1125899906842625, "label": "test_2", "properties": {}}
 {"id": 1125899906842626, "label": "test_2", "properties": {"a": 0}}
 {"id": 1125899906842627, "label": "test_2", "properties": {}}
 {"id": 1407374883553282, "label": "test_3", "properties": {}}
 {"id": 1407374883553281, "label": "test_3", "properties": {}}
 {"id": 1970324836974593, "label": "test_4", "properties": {}}
 {"id": 1970324836974594, "label": "test_4", "properties": {}}
 {"id": 2533274790395905, "label": "test_5", "properties": {}}
 {"id": 2814749767106561, "label": "test_6", "properties": {}}
 {"id": 3377699720527873, "label": "test_7", "properties": {}}
(15 rows)

MATCH (n) RETURN n;
                                  n                                  
---------------------------------------------------------------------
 {"id": 281474976710657, "label": "", "properties": {}}
 {"id": 281474976710658, "label": "", "properties": {}}
 {"id": 844424930131969, "label": "test_1", "properties": {}}
 {"id": 844424930131971, "label": "test_1", "properties": {}}
 {"id": 844424930131970, "label": "test_1", "properties": {"a": 0}}
 {"id": 1125899906842625, "label": "test_2", "properties": {}}
 {"id": 1125899906842626, "label": "test_2", "properties": {"a": 0}}
 {"id": 1125899906842627, "label": "test_2", "properties": {}}
 {"id": 1407374883553282, "label": "test_3", "properties": {}}
 {"id": 1407374883553281, "label": "test_3", "properties": {}}
 {"id": 1970324836974593, "label": "test_4", "properties": {}}
 {"id": 1970324836974594, "label": "test_4", "properties": {}}
 {"id": 2533274790395905, "label": "test_5", "properties": {}}
 {"id": 2814749767106561, "label": "test_6", "properties": {}}
 {"id": 3377699720527873, "label": "test_7", "properties": {}}
(15 rows)

CREATE ()-[:edge_multi_property { i: 5, j: 20}]->();
--
(0 rows)

MATCH ()-[e:edge_multi_property]-() RETURN e;
                                                                         e                                                                         
---------------------------------------------------------------------------------------------------------------------------------------------------
 {"id": 3659174697238529, "start_id": 281474976710659, "end_id": 281474976710660, "label": "edge_multi_property", "properties": {"i": 5, "j": 20}}
 {"id": 3659174697238529, "start_id": 281474976710659, "end_id": 281474976710660, "label": "edge_multi_property", "properties": {"i": 5, "j": 20}}
(2 rows)

MATCH ()-[e:edge_multi_property]-() REMOVE e.i, e.j RETURN e;
                                                                 e                                                                  
------------------------------------------------------------------------------------------------------------------------------------
 {"id": 3659174697238529, "start_id": 281474976710659, "end_id": 281474976710660, "label": "edge_multi_property", "properties": {}}
 {"id": 3659174697238529, "start_id": 281474976710659, "end_id": 281474976710660, "label": "edge_multi_property", "properties": {}}
(2 rows)

--Errors
REMOVE n.i;
ERROR:  syntax error at or near "REMOVE"
LINE 1: REMOVE n.i;
        ^
MATCH (n) REMOVE n.i = NULL;
ERROR:  REMOVE clause must be in the format: REMOVE variable.property_name
LINE 1: MATCH (n) REMOVE n.i = NULL;
        ^
MATCH (n) REMOVE wrong_var.i;
ERROR:  undefined reference to variable wrong_var in REMOVE clause
LINE 1: MATCH (n) REMOVE wrong_var.i;
        ^
--
-- Clean up
--
DROP FUNCTION remove_test;
ERROR:  could not find a function named "remove_test"
DROP GRAPH cypher_remove CASCADE;
NOTICE:  drop cascades to 13 other objects
DETAIL:  drop cascades to table cypher_remove._ag_label_vertex
drop cascades to table cypher_remove._ag_label_edge
drop cascades to table cypher_remove.test_1
drop cascades to table cypher_remove.test_2
drop cascades to table cypher_remove.test_3
drop cascades to table cypher_remove.test_3_edge
drop cascades to table cypher_remove.test_4
drop cascades to table cypher_remove.test_4_edge
drop cascades to table cypher_remove.test_5
drop cascades to table cypher_remove.test_6
drop cascades to table cypher_remove.e
drop cascades to table cypher_remove.test_7
drop cascades to table cypher_remove.edge_multi_property
NOTICE:  graph "cypher_remove" has been dropped
 drop_graph 
------------
 
(1 row)

--
-- End
--
